public/app/features/explore/state/query.test.ts

```diff
@@ -265,22 +273,202 @@ describe('reducer', () => {
   });
 
   describe('query rows', () => {
-    it('adds a new query row', () => {
-      reducerTester<ExploreItemState>()
-        .givenReducer(queryReducer, {
-          queries: [],
-        } as unknown as ExploreItemState)
-        .whenActionIsDispatched(
-          addQueryRowAction({
-            exploreId: ExploreId.left,
-            query: { refId: 'A', key: 'mockKey' },
-            index: 0,
-          })
-        )
-        .thenStateShouldEqual({
-          queries: [{ refId: 'A', key: 'mockKey' }],
-          queryKeys: ['mockKey-0'],
-        } as unknown as ExploreItemState);
+    describe('addQueryRowAction', () => {
+      it('should add query row when no query row yet (addQueryRowAction)', () => {
+        reducerTester<ExploreItemState>()
+          .givenReducer(queryReducer, {
+            queries: [],
+          } as unknown as ExploreItemState)
+          .whenActionIsDispatched(
+            addQueryRowAction({
+              exploreId: ExploreId.left,
+              query: { refId: 'A', key: 'mockKey' },
+              index: 0,
+            })
+          )
+          .thenStateShouldEqual({
+            queries: [{ refId: 'A', key: 'mockKey' }],
+            queryKeys: ['mockKey-0'],
+          } as unknown as ExploreItemState);
+      });
+      it('should add query row when there is already one query row (addQueryRowAction)', () => {
+        reducerTester<ExploreItemState>()
+          .givenReducer(queryReducer, {
+            queries: [{ refId: 'A', key: 'initialRow', datasource: { type: 'loki' } }],
+          } as unknown as ExploreItemState)
+          .whenActionIsDispatched(
+            addQueryRowAction({
+              exploreId: ExploreId.left,
+              query: { refId: 'B', key: 'mockKey', datasource: { type: 'loki' } },
+              index: 0,
+            })
+          )
+          .thenStateShouldEqual({
+            queries: [
+              { refId: 'A', key: 'initialRow', datasource: { type: 'loki' } },
+              { refId: 'B', key: 'mockKey', datasource: { type: 'loki' } },
+            ],
+            queryKeys: ['initialRow-0', 'mockKey-1'],
+          } as unknown as ExploreItemState);
+      });
+    });
+    describe('addQueryRow with mixed datasources disabled', () => {
+      it('should add query row when there is not yet a row', async () => {
+        const exploreId = ExploreId.left;
+        let dispatch: ThunkDispatch, getState: () => StoreState;
+
+        const store: { dispatch: ThunkDispatch; getState: () => StoreState } = configureStore({
```

Comment:
There's a lot of code duplication in the setup. Could it be extracted to a helper function for readability?

Reply:
Done.